File: day2_intcode/python/day2_intcode.py
#!/bin/python

import sys

mem = [int(n) for n in sys.argv[1].split(",")]
ip = 0
while True:
    cmd = mem[ip]
    if cmd == 1:
        a = mem[mem[ip + 1]]
        b = mem[mem[ip + 2]]
        mem[mem[ip + 3]] = a + b
        ip += 4
    elif cmd == 2:
        a = mem[mem[ip + 1]]
        b = mem[mem[ip + 2]]
        mem[mem[ip + 3]] = a * b
        ip += 4
    elif cmd == 99:
        break
    else:
        exit(1)

print(mem)        

File: day2_intcode/rust/day2_intcode.rs
use std::env;


fn main() {
    let l = env::args().skip(1).next().expect("Missing argument");
    let init_mem: Vec<i32> = l.split(",").map(|x| i32::from_str_radix(&x, 10).expect("Can't parse")).collect();
    let mut mem = init_mem.clone();
    run(&mut mem);
    println!("{}", mem[0]);
}


fn run(mem: &mut Vec<i32>) {
    let mut ip: usize = 0;
    loop {
        let cmd = mem[ip];
        match cmd {
            1 => {
                let a: i32 = mem[mem[ip + 1] as usize];
                let b: i32 = mem[mem[ip + 2] as usize];
                let dest: usize = mem[ip + 3] as usize;
                mem[dest] = a + b;
                ip += 4;
            },
            2 => {
                let a: i32 = mem[mem[ip + 1] as usize];
                let b: i32 = mem[mem[ip + 2] as usize];
                let dest: usize = mem[ip + 3] as usize;
                mem[dest] = a * b;
                ip += 4;
            },
            99 => break,
            _ => panic!("Bad opcode"),
        }
    }
}